pkg/deploy/registry/instantiate/rest.go

```diff
@@ -0,0 +1,286 @@
+// deployment config accordingly. This contains the work that the image change controller
+// had been doing up to the point we got the /instantiate endpoint.
+func (r *REST) Process(config *deployapi.DeploymentConfig, force bool) (bool, error) {
+	errs := []error{}
+	updated := false
+
+	// Process any image change triggers.
+	for _, trigger := range config.Spec.Triggers {
+		if trigger.Type != deployapi.DeploymentTriggerOnImageChange {
+			continue
+		}
+
+		params := trigger.ImageChangeParams
+
+		// Forced deployments should always try to resolve the images in the template.
+		// On the other hand, paused deployments or non-automatic triggers shouldn't.
+		if !force && (config.Spec.Paused || !params.Automatic) && len(params.LastTriggeredImage) > 0 {
+			continue
+		}
+
+		// Tag references are already validated
+		name, tag, _ := imageapi.SplitImageStreamTag(params.From.Name)
+		stream, err := r.isn.ImageStreams(params.From.Namespace).Get(name)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				errs = append(errs, err)
+			}
+			continue
+		}
+
+		// Find the latest tag event for the trigger reference.
+		latestEvent := imageapi.LatestTaggedImage(stream, tag)
+		if latestEvent == nil {
+			continue
+		}
+
+		// Ensure a change occurred
+		latestRef := latestEvent.DockerImageReference
+		if len(latestRef) == 0 || latestRef == params.LastTriggeredImage {
+			continue
+		}
+
+		// Update containers
+		names := sets.NewString(params.ContainerNames...)
+		for i := range config.Spec.Template.Spec.Containers {
+			container := &config.Spec.Template.Spec.Containers[i]
+			if !names.Has(container.Name) {
+				continue
+			}
+
+			if container.Image != latestRef {
+				// Update the image
+				container.Image = latestRef
+				// Log the last triggered image ID
+				params.LastTriggeredImage = latestRef
+				updated = true
+			}
+		}
+	}
+
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		return updated, errors.NewInternalError(err)
+	}
+
+	return updated, nil
+}
+
+// CanTrigger determines if we can trigger a new deployment for config based on the various deployment triggers.
+func (r *REST) CanTrigger(config *deployapi.DeploymentConfig, force bool) (bool, []deployapi.DeploymentCause, error) {
+	decoded, err := r.decodeFromLatest(config)
+	if err != nil {
+		return false, nil, err
+	}
+
+	ictCount, resolved, canTriggerByImageChange := 0, 0, false
+	var causes []deployapi.DeploymentCause
+
+	for _, t := range config.Spec.Triggers {
+		if t.Type != deployapi.DeploymentTriggerOnImageChange {
+			continue
+		}
+		ictCount++
+
+		// If the image is yet to be resolved then we cannot process this trigger.
+		lastTriggered := t.ImageChangeParams.LastTriggeredImage
+		if len(lastTriggered) == 0 {
+			continue
+		}
+		resolved++
+
+		// Non-automatic triggers should not be able to trigger deployments.
+		if !t.ImageChangeParams.Automatic {
+			continue
+		}
+
+		// We need stronger checks in order to validate that this template
```

Comment:
I don't think I handle the case where the annotation is not present? I am returning the self only when latestVersion=0 which makes sense I think.